cs/src/core/Transcoder.cs

```diff
@@ -93,16 +124,79 @@ public Transcoder()
         /// <param name="writer">Writer instance</param>
         public void Transcode(R reader, W writer)
         {
-            transcode[0](reader, writer);
+            helper.Transcode(reader, writer);
+        }
+
+        private class TranscoderHelper
+        {
+            readonly Action<R, W>[] transcode;
+
+            public TranscoderHelper(RuntimeSchema schema)
+            {
+                transcode = Generate(schema);
+            }
+
+            public TranscoderHelper(Type type)
+            {
+                transcode = Generate(type);
+            }
+
+            virtual public void Transcode(R reader, W writer)
+            {
+                transcode[0](reader, writer);
+            }
+
+            private Action<R, W>[] Generate<S>(S schema)
+            {
+                var parser = ParserFactory<R>.Create(schema);
+                return SerializerGeneratorFactory<R, W>.Create(
+                        (r, w, i) => transcode[i](r, w), schema)
+                    .Generate(parser)
+                    .Select(lambda => lambda.Compile()).ToArray();
+            }
         }
 
-        Action<R, W>[] Generate<S>(S schema)
+        private class TwoPassTranscoderHelper<FPW> : TranscoderHelper
         {
-            var parser = ParserFactory<R>.Create(schema);
-            return SerializerGeneratorFactory<R, W>.Create(
-                    (r, w, i) => transcode[i](r, w), schema)
-                .Generate(parser)
-                .Select(lambda => lambda.Compile()).ToArray();
+            readonly Action<R, FPW>[] firstPassTranscode;
+
+            public TwoPassTranscoderHelper(RuntimeSchema schema):
+                base(schema)
+            {
+                firstPassTranscode = GenerateFirstPass(schema);
+            }
+
+            public TwoPassTranscoderHelper(Type type):
+                base(type)
+            {
+                firstPassTranscode = GenerateFirstPass(type);
+            }
+
+            public override void Transcode(R reader, W writer)
+            {
+                var firstPassWriter = ((ITwoPassProtocolWriter)writer).GetFirstPassWriter();
+                if (firstPassWriter != null)
+                {
+                    if (!typeof(ICloneable<R>).IsAssignableFrom(typeof(R)))
```

Comment:
We don't know whether the CompactBinaryWriter passed in for W is using v1 or v2 until we get to here, so we may not need R to be ICloneable.